### PR TITLE
Add dropout operator.

### LINKED ISSUE #3816: Add dropout operator.

### PR DESCRIPTION
**Pull request by xinghai-sun**
Resolve #3816

### CODE REVIEW

**Comment by xinghai-sun** [comment]
Dropout behaves differently in train and test mode: e.g. in the test mode, the dropout is often turned off. How can the dropout operator know whether it's in the test mode or the train mode?

File: paddle/operators/dropout_op.cc [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    ctx.Output<Tensor>("Mask")->Resize(dims);
+  }
+};
+
+class DropoutOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  DropoutOpMaker(framework::OpProto *proto,
+                 framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddAttr<float>("dropout_prob", "Dropout probability.").SetDefault(.5f);
```
Comment (qingqing01) [inline paddle/operators/dropout_op.cc]: need more detail comments, 是以概率probability置0，还是以(1-probability)置0，这块需要解释清楚。

File: paddle/operators/dropout_op.cc [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+class DropoutOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("X"), "Input(X) must not be null.");
+    auto dims = ctx.Input<Tensor>("X")->dims();
+    ctx.Output<Tensor>("Out")->Resize(dims);
+    ctx.Output<Tensor>("Mask")->Resize(dims);
```
Comment (qingqing01) [inline paddle/operators/dropout_op.cc]: 需要检查属性`context.GetAttr<float>("dropout_prob");` 大于0

File: paddle/operators/dropout_op.h [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    engine.seed(seed);
+    std::uniform_real_distribution<T> dist(0, 1);
+    size_t size = framework::product(mask->dims());
+    for (size_t i = 0; i < size; ++i) {
+      if (dist(engine) < dropout_prob) {
+        mask_data[i] = 0;
+        y_data[i] = 0;
+      } else {
+        mask_data[i] = 1;
+        y_data[i] = (1 - dropout_prob) * x_data[i];
```
Comment (qingqing01) [inline paddle/operators/dropout_op.h]: 对于training应该是：y_data[i] =x_data[i]

File: paddle/operators/dropout_op.h [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* x = context.Input<Tensor>("X");
+    auto* y = context.Output<Tensor>("Out");
+    auto* mask = context.Output<Tensor>("Mask");
+    T* mask_data = mask->mutable_data<T>(context.GetPlace());
+    T* y_data = y->mutable_data<T>(context.GetPlace());
+    const T* x_data = x->data<T>();
+
+    float dropout_prob = context.op_.GetAttr<float>("dropout_prob");
+    int seed = context.op_.GetAttr<int>("seed");
```
Comment (qingqing01) [inline paddle/operators/dropout_op.h]: `context.op_.GetAttr` -> `context.GetAttr`

File: paddle/operators/dropout_op.h [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+class GPUDropoutKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* x = context.Input<Tensor>("X");
+    auto* y = context.Output<Tensor>("Out");
+    auto* mask = context.Output<Tensor>("Mask");
+    y->mutable_data<T>(context.GetPlace());
+
+    float dropout_prob = context.op_.GetAttr<float>("dropout_prob");
+    int seed = context.op_.GetAttr<int>("seed");
```
Comment (qingqing01) [inline paddle/operators/dropout_op.h]: ` context.op_.GetAttr` ->  `context.GetAttr`

File: paddle/operators/dropout_op.h [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        y_data[i] = (1 - dropout_prob) * x_data[i];
+      }
+    }
+  }
+};
+
+template <typename T>
+struct MaskGenerator {
+  float dropout_prob_;
+  int seed_;
```
Comment (qingqing01) [inline paddle/operators/dropout_op.h]: There is no `_` for the data member of struct

https://google.github.io/styleguide/cppguide.html#Variable_Names

File: paddle/operators/dropout_op.h (hunk 5 above)
Comment (xinghai-sun) [reply paddle/operators/dropout_op.h]: Done.

File: paddle/operators/dropout_op.h (hunk 4 above)
Comment (xinghai-sun) [reply paddle/operators/dropout_op.h]: Done.

File: paddle/operators/dropout_op.cc [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  }
+};
+
+class DropoutOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  DropoutOpMaker(framework::OpProto *proto,
+                 framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddAttr<float>("dropout_prob", "Probability for dropping out units.")
```
Comment (qingqing01) [inline paddle/operators/dropout_op.cc]: The template type is needed for attr.
```c++
template <typename AttrType>
class DropoutOpMaker : public framework::OpProtoAndCheckerMaker {
 public:
  DropoutOpMaker(framework::OpProto *proto,
                  framework::OpAttrChecker *op_checker)
       : OpProtoAndCheckerMaker(proto, op_checker) {
     AddAttr<AttrType>("dropout_prob", "Probability for dropping out units.")
```

File: paddle/operators/dropout_op.h [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+using EigenMatrix = framework::EigenMatrix<T, MajorType, IndexType>;
+
+template <typename Place, typename T>
+class CPUDropoutKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* x = context.Input<Tensor>("X");
+    auto* y = context.Output<Tensor>("Out");
+    auto* mask = context.Output<Tensor>("Mask");
+    T* mask_data = mask->mutable_data<T>(context.GetPlace());
```
Comment (qingqing01) [inline paddle/operators/dropout_op.h]: ```mask->mutable_data<int>```

The mask is int type.

File: paddle/operators/dropout_op.h [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    T* y_data = y->mutable_data<T>(context.GetPlace());
+    const T* x_data = x->data<T>();
+
+    float dropout_prob = context.GetAttr<float>("dropout_prob");
```
Comment (qingqing01) [inline paddle/operators/dropout_op.h]: template type is needed for attr.

File: paddle/operators/dropout_op.h [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    auto* x = context.Input<Tensor>("X");
+    auto* y = context.Output<Tensor>("Out");
+    auto* mask = context.Output<Tensor>("Mask");
+    y->mutable_data<T>(context.GetPlace());
+
+    float dropout_prob = context.GetAttr<float>("dropout_prob");
+    int seed = context.GetAttr<int>("seed");
+    thrust::counting_iterator<unsigned int> index_sequence_begin(0);
+    int size = framework::product(mask->dims());
+    T* mask_data = mask->mutable_data<T>(context.GetPlace());
```
Comment (qingqing01) [inline paddle/operators/dropout_op.h]: ```mask->mutable_data<int>```

File: paddle/operators/dropout_op.h [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    thrust::transform(index_sequence_begin, index_sequence_begin + size,
+                      thrust::device_ptr<T>(mask_data),
+                      MaskGenerator<T>(dropout_prob, seed));
```
Comment (qingqing01) [inline paddle/operators/dropout_op.h]: Maybe the CPUKernel can be implemented with same way, `std::transform` can be used.

File: paddle/operators/dropout_op.h [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* grad_x = context.Output<Tensor>(framework::GradVarName("X"));
+    auto* grad_y = context.Input<Tensor>(framework::GradVarName("Out"));
+    auto* mask = context.Input<Tensor>("Mask");
+    grad_x->mutable_data<T>(context.GetPlace());
+
+    auto dims = grad_x->dims();
+    int size = static_cast<int>(framework::product(dims));
+    auto new_dims = framework::make_ddim({dims[0], size / dims[0]});
+    auto M = EigenMatrix<T>::From(*mask, new_dims);
```
Comment (qingqing01) [inline paddle/operators/dropout_op.h]: ```EigenMatrix<int>```

File: paddle/operators/dropout_op.cc [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    // validity check
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("X"), "Input(X) must not be null.");
+    PADDLE_ENFORCE_GE(ctx.GetAttr<float>("dropout_prob"), 0);
+    PADDLE_ENFORCE_LE(ctx.GetAttr<float>("dropout_prob"), 1);
+    // resize
+    auto dims = ctx.Input<Tensor>("X")->dims();
+    ctx.Output<Tensor>("Out")->Resize(dims);
+    ctx.Output<Tensor>("Mask")->Resize(dims);
```
Comment (qingqing01) [inline paddle/operators/dropout_op.cc]: LoDTensor合入后， InferShape里的Output需要改成：`Output< framework::LoDTensor>`

File: paddle/operators/dropout_op.cc [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+    PADDLE_ENFORCE_LE(ctx.GetAttr<float>("dropout_prob"), 1);
+    auto x_dims = ctx.Input<Tensor>("X")->dims();
+    auto mask_dims = ctx.Input<Tensor>("Mask")->dims();
+    auto out_dims = ctx.Input<Tensor>(framework::GradVarName("Out"))->dims();
+    PADDLE_ENFORCE_EQ(x_dims, out_dims,
+                      "Dimensions of Input(X) and Out@Grad must be the same.");
+    PADDLE_ENFORCE_EQ(x_dims, mask_dims,
+                      "Dimensions of Input(X) and Mask must be the same.");
+    // resize
+    auto *x_grad = ctx.Output<Tensor>(framework::GradVarName("X"));
```
Comment (qingqing01) [inline paddle/operators/dropout_op.cc]: LoDTensor合入后， InferShape里的Output需要改成：`Output< framework::LoDTensor>`

File: python/paddle/v2/framework/tests/test_dropout_op.py [hunk 15]
```diff
@@ -0,0 +1,56 @@
+import unittest
+import numpy as np
+from gradient_checker import GradientChecker, create_op
+from op_test_util import OpTestMeta
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_dropout_op.py]: 需要使用新的单测框架

File: paddle/operators/dropout_op.cc (hunk 13 above)
Comment (xinghai-sun) [reply paddle/operators/dropout_op.cc]: Done.

File: paddle/operators/dropout_op.cc (hunk 14 above)
Comment (xinghai-sun) [reply paddle/operators/dropout_op.cc]: Done.

File: python/paddle/v2/framework/tests/test_dropout_op.py (hunk 15 above)
Comment (xinghai-sun) [reply python/paddle/v2/framework/tests/test_dropout_op.py]: Done.

File: paddle/operators/dropout_op.h (hunk 12 above)
Comment (xinghai-sun) [reply paddle/operators/dropout_op.h]: I think any of `int`, `float`, or `T` for mask data type is OK. `T` might be better because it avoids implicit type conversion when mask is multiplied with X.

When considering memory usage, it would be better to use a `bool` (single byte) or `float16` (2 bytes), but that will incur the following error:

```
129: .terminate called after throwing an instance of 'paddle::platform::EnforceNotMet'
129:   what():  This type of tensor cannot be expose to Python at [/home/work/sunxinghai/git/Paddle/paddle/pybind/tensor_py.h:36]
```
What do you think?

File: paddle/operators/dropout_op.cc (hunk 7 above)
Comment (xinghai-sun) [reply paddle/operators/dropout_op.cc]: I think for attributes like dropout_prob, learning_rate, etc.., we can fix them to `float`, instead of of a flexible `AttrType` (I know it might be `float`, or `double`, but it is also dangerous if set to `int` or others) . I don't see any necessity in using such a typename.

File: paddle/operators/dropout_op.h (hunk 9 above)
Comment (xinghai-sun) [reply paddle/operators/dropout_op.h]: As discussed above.

File: paddle/operators/dropout_op.h (hunk 12 above)
Comment (qingqing01) [reply paddle/operators/dropout_op.h]: Sorry, my comment is wrong. Just use type T.  Since the tensor in pybind doest not support to set and get tensor with bool type, the error for `bool` type is caused.
